eng/versioning/update_versions.py

```diff
@@ -0,0 +1,194 @@
+                    if version_type == 'current':
+                        try:
+                            new_version = module.current
+                        except AttributeError:
+                            raise ValueError('Module: {0} does not have a current version.\nFile={1}\nLine={2}'.format(module_name, target_file, line))
+                    elif version_type == 'dependency':
+                        new_version = module.dependency
+                    else:
+                        raise ValueError('Invalid version type: {} for module: {}.\nFile={}\nLine={}'.format(version_type, module_name, target_file, line))
+
+                    newline = re.sub(version_regex_str, new_version, line)
+                    newlines.append(newline)
+                    file_changed = True
+                else:
+                    newlines.append(line)
+
+                if not repl_open:
+                    module_name, version_type = '', ''
+
+        if file_changed:
+            with open(target_file, 'w', encoding='utf-8') as f:
+                for line in newlines:
+                    f.write(line)
+    except Exception as e:
+        print("Unexpected exception: " + str(e))
+
+def load_version_map_from_file(the_file, version_map):
+    with open(the_file) as f:
+        for raw_line in f:
+            stripped_line = raw_line.strip()
+            if not stripped_line or stripped_line.startswith('#'):
+                continue
+            module = CodeModule(stripped_line)
+            version_map[module.name] = module
+
+def display_version_info(version_map):
+    for value in version_map.values():
+        print(value)
+    
+def update_versions_all(update_type, build_type, target_file):
+    version_map = {}
+    # Load the version and/or external dependency file for the given UpdateType
+    # into the verion_map. If UpdateType.all is selected then versions for both
+    # the libraries and external dependencies are being updated.
+    if update_type == UpdateType.library or update_type == UpdateType.all:
+        version_file = os.path.normpath('eng/versioning/version_' + build_type.name + '.txt')
+        print('version_file=' + version_file)
+        load_version_map_from_file(version_file, version_map)
+
+    if update_type == UpdateType.external_dependency or update_type == UpdateType.all:
+        dependency_file = os.path.normpath('eng/versioning/external_dependency_' + build_type.name + '.txt')
+        print('external_dependency_file=' + dependency_file)
+        load_version_map_from_file(dependency_file, version_map)
+
+    display_version_info(version_map)
+
+    if target_file:
+        update_versions(version_map, target_file)
+    else:
+        for root, _, files in os.walk("."):
+            for file_name in files:
+                file_path = root + os.sep + file_name
+                if file_name == 'README.md' or (file_name.startswith('pom.') and file_name.endswith('.xml')):
```

Comment:
Is there a reason to limit to README.md as opposed to all md files? I'm not sure if there are any other cases but I can imagine samples potentially going in other md files. Probably OK for now but something to consider for later.

Reply:
Good call. I've checked the entire source base and there are other .md files with versions but they're not in data or client, only management. I'll update this to include all .md files when I do the changes for external dependencies.